File: day53/minDepth.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# recursion
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root: return 0
        
        if not root.left: return 1 + self.minDepth(root.right)
        if not root.right: return 1 + self.minDepth(root.left)
        
        return 1 + min(self.minDepth(root.left), self.minDepth(root.right))

# class Solution:
#     def minDepth(self, root: TreeNode) -> int:
#         if not root: return 0
#         left = self.minDepth(root.left)
#         right = self.minDepth(root.right)
#         return 1 + left + right if left * right == 0 else 1 + min(left, right)

# BFS
# class Solution:
#     def minDepth(self, root: TreeNode) -> int:
#         if not root: return 0

#         res = 0
#         queue = collections.deque()
#         queue.append(root)

#         while queue:
#             res += 1
#             size = len(queue)
#             for _ in range(size):
#                 cur = queue.popleft()
#                 if not cur.left and not cur.right: return res
#                 if cur.left: queue.append(cur.left)
#                 if cur.right: queue.append(cur.right)

File: day53/minDepth.java
/*
111. 二叉树的最小深度
给定一个二叉树，找出其最小深度。
最小深度是从根节点到最近叶子节点的最短路径上的节点数量。
说明: 叶子节点是指没有子节点的节点。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/minimum-depth-of-binary-tree
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;

        int left = minDepth(root.left);
        int right = minDepth(root.right);

        return left * right == 0 ? 1 + left + right : 1 + Math.min(left, right);
    }
}

// class Solution {
//     public int minDepth(TreeNode root) {
//         if (root == null) return 0;

//         if (root.left == null) return 1 + minDepth(root.right);
//         if (root.right == null) return 1 + minDepth(root.left);

//         return 1 + Math.min(minDepth(root.left), minDepth(root.right));

//     }
// }
